autogpt_platform/backend/backend/server/v2/AutoMod/manager.py

```diff
@@ -0,0 +1,296 @@
+                        ExecutionStatus.INCOMPLETE,
+                    ],
+                )
+                self.handle_moderation_failure(
+                    db_client=db_client,
+                    executions=existing_executions,
+                    send_update_func=send_update_func,
+                )
+                return False, Exception("Execution failed due to content moderation")
+
+            return True, None
+
+        except Exception as e:
+            logger.error(f"Input moderation execution failed: {e}")
+            return False, Exception("Execution failed due to content moderation")
+
+    def moderate_graph_execution_outputs(
+        self,
+        db_client,
+        graph_exec_id: str,
+        user_id: str,
+        graph_id: str,
+        event_loop,
+        send_update_func,
+        timeout: int = 10,
+    ) -> Tuple[bool, Exception | None]:
+        """
+        Complete output moderation flow for graph execution
+        Returns: (success, error_if_failed)
+        """
+        if not self.config.moderate_outputs or not self.config.enabled:
+            return True, None
+
+        # Get completed executions and collect outputs
+        completed_executions = db_client.get_node_executions(
+            graph_exec_id, statuses=[ExecutionStatus.COMPLETED], include_exec_data=True
+        )
+
+        if not completed_executions:
+            return True, None
+
+        all_outputs = []
+        for exec_entry in completed_executions:
+            if exec_entry.output_data:
+                all_outputs.extend(str(v) for v in exec_entry.output_data.values() if v)
+
+        if not all_outputs:
+            return True, None
+
+        # Combine all content and moderate directly
+        content = " ".join(all_outputs)
+
+        # Run moderation
+        try:
+            moderation_passed = asyncio.run_coroutine_threadsafe(
+                self._moderate_content(
+                    content,
+                    {
+                        "user_id": user_id,
+                        "graph_id": graph_id,
+                        "graph_exec_id": graph_exec_id,
+                        "moderation_type": "execution_output",
+                    },
+                ),
+                event_loop,
+            ).result(timeout=timeout)
+
+            if not moderation_passed:
+                self.handle_moderation_failure(
+                    db_client=db_client,
+                    executions=completed_executions,
+                    send_update_func=send_update_func,
+                )
+                return False, Exception("Execution failed due to content moderation")
+
+            return True, None
+
+        except Exception as e:
+            logger.error(f"Output moderation execution failed: {e}")
+            return False, Exception("Execution failed due to content moderation")
+
+    def handle_moderation_failure(self, db_client, executions: List, send_update_func):
```

Comment:
Why do we need to do this? If we clear **all** input & output anyway, why don't we just mark the error and graph level, and not return any input & output when the graph is marked as failing moderation?

This way, we can't avoid updating the db entirely. Plus, no need to call `send_update_func` since raising an Error will do this. (just make sure to update execution_stats before raising the error so it will get updated.